### PR TITLE
Add IMU data

### PR DESCRIPTION
**Pull request by realbigmitchell**
I added listeners for the accelerometer, gyroscope, magnetometer, pressure, rotation vector, and linear acceleration sensors. This provides users with the same data used in Google Play's Fused Location Provider so they can use these additional readings in combination with raw GNSS data. The IMU update rate is too high to output values to the UI logger for longer than a few seconds without bogging it down, so I only implemented the onSensorChanged method for the file logger.

### CODE REVIEW

**Review by sschava1 — APPROVED** [review]

**Review by sschava1 — APPROVED** [review]

**Comment by alexanderkuznetsov96** [comment]
Mitchell, could you please resolve these conflicts? Thanks!

**Comment by realbigmitchell** [comment]
> Mitchell, could you please resolve these conflicts? Thanks!

Done! I also addressed your outstanding requests from my prior PR.

File: GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/FileLogger.java [hunk 1]
```diff
@@ -339,6 +366,25 @@ public void onNmeaReceived(long timestamp, String s) {
     }
   }
 
+  @Override
+  public void onSensorChanged(SensorEvent event) {
```
Comment (alexanderkuznetsov96) [inline GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/FileLogger.java:370]: Please add a javadoc to this method.

File: GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/FileLogger.java [hunk 2]
```diff
@@ -339,6 +366,25 @@ public void onNmeaReceived(long timestamp, String s) {
...
+    synchronized (mFileLock) {
+      if(mFileWriter == null) {
+        return;
+      }
+      mStringBuilder.setLength(0);
+      mStringBuilder.append(event.sensor.getStringType().substring(event.sensor.getStringType().lastIndexOf('.') + 1)).append(RECORD_DELIMITER);
```
Comment (alexanderkuznetsov96) [inline GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/FileLogger.java:376]: I recommend handling each sensor type separately using a series of `if else` blocks. I.e.

```
if (sensorEvent.sensor.getType() == Sensor.TYPE_ACCELEROMETER_UNCALIBRATED) { 
...
} else if ( ....
```

File: GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/FileLogger.java [hunk 3]
```diff
@@ -339,6 +366,25 @@ public void onNmeaReceived(long timestamp, String s) {
...
+      mStringBuilder.append(event.timestamp).append(RECORD_DELIMITER);
+      mStringBuilder.append(java.util.Arrays.toString(event.values).replaceAll("[] ]", "").replace("[", ""));
```
Comment (alexanderkuznetsov96) [inline GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/FileLogger.java:378]: This is some slick code golf, but it's a bit hard to follow. I recommend replacing it with something simpler to read. Something like:

```
String.format(
            "Accel,%s,%s,%s,%s,%s,%s,%s",
            event.timestamp,
            event.values[IMU_X_INDEX], 
            event.values[IMU_Y_INDEX],
            event.values[IMU_Z_INDEX], 
            event.values[IMU_X_BIAS_INDEX], 
            event.values[IMU_Y_BIAS_INDEX], 
            event.values[IMU_Z_BIAS_INDEX]);
```
You can extract that into a helper method, and for each sensor, just pass a different sensor type string. The rest seems common.

File: GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/FileLogger.java [hunk 4]
```diff
@@ -163,6 +165,31 @@ public void startNewLog() {
         currentFileWriter.newLine();
         currentFileWriter.write(COMMENT_START);
         currentFileWriter.newLine();
+        currentFileWriter.write(COMMENT_START);
+        currentFileWriter.write("accelerometer,timestamp,x,y,z");
+        currentFileWriter.newLine();
+        currentFileWriter.write(COMMENT_START);
+        currentFileWriter.newLine();
+        currentFileWriter.write(COMMENT_START);
+        currentFileWriter.write("gyroscope,timestamp,x,y,z");
```
Comment (alexanderkuznetsov96) [inline GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/FileLogger.java:174]: These new headers should match the formatting of the others: first letter capitalized. Also, be sure to include units in all the new headers. For example timestampMillis (if it's in milliseconds, nanos if it's in nanoseconds, etc) and the measurement values themselves can be [X,Y,Z]GyroRateRadPerSec, respectively.

File: GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/SettingsFragment.java [hunk 5]
```diff
@@ -195,15 +195,36 @@ public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
...
+                          mMeasurementProvider.registerSensors();
+                      } else {
+                          registerImuLabel.setText("Switch is OFF");
+                          mMeasurementProvider.unregisterSensors();
+                      }
+                  }
+              });
+
+
+      final Switch autoScroll = (Switch) view.findViewById(R.id.auto_scroll_on);
```
Comment (alexanderkuznetsov96) [inline GNSSLogger/app/src/main/java/com/google/android/apps/location/gps/gnsslogger/SettingsFragment.java:227]: This block looks out of line. Check the indenting?

**Review by alexanderkuznetsov96 — CHANGES_REQUESTED** [review]

**Comment by realbigmitchell** [comment]
I'm a bit confused. I noticed after submitting this PR that the GnssLogger app had been added to Google Play on November 16th and updated to version 3.0. It looks like the new version has several new features, including logging for both the Fused Location Provider and sensor readings, but this Github repo has not been updated to reflect those changes.

I also noticed that you all recently posted data from the GnssLogger app on Kaggle which includes sensor and FLP data. Some of the files (e.g. 2020-09-04-US-SF-1\Pixel4_GnssLog.txt) have uncalibrated sensor data, while others (e.g. 2020-05-14-US-MTV-1\Pixel4_GnssLog.txt) have only data with the calibration already applied.

Has Google decided not to continue sharing its work on the source code for this app? If so, why? And if not, would it make more sense for you all to update the Github repo to the latest version which already has this feature implemented?

**Comment by guoyufu** [comment]
Hello realbigmitchell, thanks for your PR and your interest in GnssLogger and the released datasets!

Since this repo was created 4 years ago, the aim is set to provide the community an example source code to process Android raw GNSS measurement APIs. Further opensource development is not in our plan, so as to keep this baseline example in a good shape. However, we welcome small and fine community contributions such as this one to improve this main repo and benefit the rest users. 

GnssLogger at PlayStore is another thread of our internal work. We are polishing it along the way. The "Calibrated/Uncalibrated"  confusion was a bug in the summer of 2020. Please see https://www.kaggle.com/google/android-smartphones-high-accuracy-datasets/discussion/209232 for my explanation about it. 

PR, comments, reviews to the datasets, this repo, and the GnssLogger at PlayStore are appreciated and valued by us. Thanks for your support to our works!

**Comment by johnsonmitchelld** [comment]
Thanks for the explanation @guoyufu and PR feedback @skzv.

I believe I have addressed all of Sasha's feedback at this point. All of the sensor listeners and log file headers have been modified to match those used in the last dataset provided on Kaggle, 2020-09-04-US-SF-12. The only difference is the timestamp implementation, which I tried to keep as similar to the raw GNSS measurement implementation as possible by using the sensor timestamp value.

Let me know if you have any more suggestions!

**Comment by barbeau** [comment]
(an aside - sorry, I'd open an issue on this, but issues aren't allowed on this repo)

@guoyufu Thanks for the explanation about this project vs. the Play store version of GnssLogger, it's great to know the future roadmap of the two projects. I'd suggest adding a note on this divergence to the README of this repo, just to make this distinction clear to the community. There may be subtle differences between implementations (e.g., this PR) that could result in slightly different datasets, and I could easily see someone sinking hours of time in to trying to figure out why data they collected via the Play Store version doesn't align exactly with data they or someone else collected via this OSS version (thinking they are from the same source code).

**Comment by mash-graz** [comment]
dear gps-measurement maintainers!

when i found the GNSS Raw Measurements Taskforce Workshop 2020 reports and read about the coming improvements in the Raw GNSS Logger in Frank van Diggelens presentation, i was really fascinated and eagerly looking forward. it's therefore a little bit disappointing, taht this changes will not be pubished as open source code anymore.

well -- to some degree i'm not really surprised. this kind of policy changes are unfortunately happening much to often. so it's just another example to refer, when younger colleagues don't understand, why stupid diehards like me still prefer the rather strict rules of the GPL to protect open source code projects from this kind of unforeseeable disappearance resp. split between an outdated/incomplete open source branch and a more attractive closed source product. nevertheless it's just a sad development, which most likely nobody of us can change.

i really have to confim @barbeau s claim:
>There may be subtle differences between implementations (e.g., this PR) that could result in slightly different datasets, and I could easily see someone sinking hours of time in to trying to figure out why data they collected...

i'm still developing a set of open source tools to parse GNSS Logger data and transform it to Apache Arrow dataframes resp. parquet/feather storage for much more efficient post-processing and programing language agnostic type safe data exchange. while working on this project i already stumbled over a bunch of very unpleasant subtle incompatibilities and ambiguities in the output of different GNSS Logger file format implementations. 

i definitely would highly appreciate one feature complete canonical open source reference implementation and shared space for feedback and open discussion of development questions concerning the android GNSS measurement infrastructure.

> (an aside - sorry, I'd open an issue on this, but issues aren't allowed on this repo)

yes -- github issues would be very helpful to get closer to this already mentioned desires.

but because they are not available until now, i want to take the chance and express at least two additional points, which i would otherwise submit as individual feature requests or bug reports:

* because i usually prefer to write code in the rust programming language, wherever possible, to maximize the performance and correctness, i had to face the problem, that GNSS measurements are one of those interfaces on the android platform, which are not accessible via NDK. that's a rather unsatisfying restriction, because native code makes a lot of sense for fast and energy efficient processing of complex GNSS data conditioning. maybe someone here can give me a hint, how this could be already archived by some undocumented loopholes or what's already in the works to solve this demand in the long run? -- i simply couldn't figure out any nice solution for myself until now.

* it's really a pity, that some grave flaws and obvious bugs regarding the actual RAW GNSS data provided by existing android devices are still limiting the practical relevance and attractivity of this kind of capabilities.

as a very realistic example, just take a closer look at the Xiaomi MI 8 results provided in your  Smartphone Decimeter Challenge dataset:

Bildschirmfoto_2021-01-13_23-29-34

the results -- they all look very similar to the red line shown in this screenshot -- are indeed fascinating, because they are utterly unsatisfying and more or less useless in practice.

well -- this may may be seen a funny joke or sideswipe against some competing manufacturer and a rather popular mobile device mentioned in many recent GNSS research papers in particular, but it also undermines the common goal of attractivity and practical usefulness of more advanced GNSS RAW data processing on the android platform in general! perhaps it would be therefore more adequate, to identify and solve the actual cause of this obvious issues and more than insufficient behavior resp. results.

sorry for somehow hijacking this PR, but the policy changes resp. closed source future of the  GNSSLogger app looks like a rather significant and disappointing development to me.

**Review by skzv — APPROVED** [review]

**Comment by skzv** [comment]
Thanks for your contributions @johnsonmitchelld!
